apollo-federation/src/connectors/json_selection/apply_to.rs

```diff
@@ -5135,4 +5120,424 @@ mod tests {
+        );
+    }
+
+    #[test]
+    fn question_operator_preserves_non_null_shapes() {
+        let spec = ConnectSpec::V0_3;
+
+        let shape_context = {
+            let mut named_shapes = IndexMap::default();
+
+            named_shapes.insert(
+                "$root".to_string(),
+                Shape::record(
+                    {
+                        let mut map = Shape::empty_map();
+                        map.insert("nonNullString".to_string(), Shape::string([]));
+                        map
+                    },
+                    [],
+                ),
+            );
+
+            ShapeContext::new(SourceId::Other("JSONSelection".into()))
+                .with_spec(spec)
+                .with_named_shapes(named_shapes)
+        };
+
+        let non_null_selection = selection!("$(nonNullString?)", spec);
+
+        assert_eq!(
+            non_null_selection
+                .compute_output_shape(
+                    &shape_context,
+                    shape_context.named_shapes().get("$root").unwrap().clone(),
+                )
+                .pretty_print(),
+            "String",
+        );
+    }
+
+    #[test]
+    fn question_operator_with_multiple_operators_in_chain() {
+        let spec = ConnectSpec::V0_3;
+
+        // Test combining ? with other operators
+        let mixed_chain_selection = selection!("$(field? ?? 'default')", spec);
+        assert_eq!(
+            mixed_chain_selection.apply_to(&json!({"field": "value"})),
+            (Some(json!("value")), vec![]),
+        );
+        assert_eq!(
+            mixed_chain_selection.apply_to(&json!({"field": null})),
+            (Some(json!("default")), vec![]),
+        );
+        assert_eq!(
+            mixed_chain_selection.apply_to(&json!({})),
+            (Some(json!("default")), vec![]),
+        );
+    }
+
+    #[test]
+    fn question_operator_direct_null_input_shape() {
+        let spec = ConnectSpec::V0_3;
+
+        let shape_context = {
+            let mut named_shapes = IndexMap::default();
+
+            named_shapes.insert("$root".to_string(), Shape::null([]));
+
+            ShapeContext::new(SourceId::Other("JSONSelection".into()))
+                .with_spec(spec)
+                .with_named_shapes(named_shapes)
+        };
+
+        let null_selection = selection!("$root?", spec);
+
+        assert_eq!(
+            null_selection
+                .compute_output_shape(
+                    &shape_context,
+                    shape_context.named_shapes().get("$root").unwrap().clone(),
+                )
+                .pretty_print(),
+            "None",
+        );
+    }
+
+    #[test]
+    fn test_unknown_name() {
+        let spec = ConnectSpec::V0_3;
+        let sel = selection!("book.author? { name age? }", spec);
+        assert_eq!(
+            sel.shape().pretty_print(),
+            "One<{ age: $root.book.author?.*.age?, name: $root.book.author?.*.name }, None>",
+        );
+    }
```

Comment:
Here's an example of a now-passing test like the one I hoped for in https://github.com/apollographql/router/pull/8330#discussion_r2382577465.

This works because we are now representing the `?` operator as part of names like `$root.book.author?.*.age?`, as discussed.